src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs

```diff
@@ -47,79 +49,61 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+                // TODO: this should have cancellation token
+                await status.WaitForCompletionAsync().ConfigureAwait(false);
             }
 
-            public Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
+            public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
             {
-                if (_workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_ForceDelay))
+                var status = await GetProgressStageStatusAsync().ConfigureAwait(false);
+                if (status == null)
                 {
-                    // this is for prototype/mock for people/teams trying partial solution load prototyping
-                    // it shouldn't concern anyone outside of that group. 
-                    //
-                    // "experimentationService.IsExperimentEnabled(WellKnownExperimentNames.PartialLoadMode)" check above
-                    // make sure this part of code "Service : IWorkspaceStatusService" is not exposed anyone outside of
-                    // the partial solution load group.
-                    //
-                    // for ones that is in the group, one can try lightbulb behavior through internal option page
-                    // such as moving around caret on lines where LB should show up.
-                    //
-                    // it works as below
-                    // 1. when this is first called, we return false and start timer to change its status in delayInMS
-                    // 2. once delayInMs passed, we clear the delay and return true.
-                    // 3. if it is called before the timeout, we reset the timer and return false
-                    if (_lastTimeCalled == null)
-                    {
-                        var delay = _workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_DelayInMS);
-                        _lastTimeCalled = new ResettableDelay(delayInMilliseconds: delay, AsynchronousOperationListenerProvider.NullListener);
-                        _ = _lastTimeCalled.Task.SafeContinueWith(_ => StatusChanged?.Invoke(this, true), TaskScheduler.Default);
-
-                        return SpecializedTasks.False;
-                    }
+                    return false;
+                }
 
-                    if (_lastTimeCalled.Task.IsCompleted)
-                    {
-                        _lastTimeCalled = null;
-                        return SpecializedTasks.True;
-                    }
+                return !status.Status.IsInProgress;
+            }
 
-                    _lastTimeCalled.Reset();
-                    return SpecializedTasks.False;
+            private async Task<IVsOperationProgressStageStatus> GetProgressStageStatusAsync()
+            {
+                var service = await _serviceProvider.GetServiceAsync<SVsOperationProgress, IVsOperationProgressStatusService>(throwOnFailure: false).ConfigureAwait(false);
```

Comment:
is there a GetServiceAsync that takes a cancellation token?

Reply:
ya. I looked and there isn't one. so I guess they don't expect it to be cancellable. @AArnott

Reply:
It isn't. Getting services (and loading packages) are fundamentally not cancelable operations. Their result (whether successful or I think unsuccessful) is stored and replayed for every subsequent call. Therefore once the first caller expresses interest, we are bound to complete that work since aborting it would screw up the state of that service/package for all subsequent callers. 
So if you want to cancel, it's purely for your own use so you can resume faster. And for that, we have the `Task.WithCancellation(CancellationToken)` extension method.

Reply:
Ah, that would be nice to use.  Thanks @AArnott .